### ISSUE TITLE
test

### ISSUE THREAD

**Issue by OnthewayLsd** [OnthewayLsd/test#1]
# -*- coding: utf-8 -*-
import tensorflow as tf
import numpy as np
import random
import time
import sys

#from game_state import GameState
#from game_state import ACTION_SIZE

from game_state_2 import GameState

from game_ac_network_2 import GameACFFNetwork, GameACLSTMNetwork

import constants_2 as constants
from constants_2 import GAMMA
from constants_2 import LOCAL_T_MAX
from constants_2 import ENTROPY_BETA
from constants_2 import USE_LSTM
from constants_2 import ACTION_SIZE
from constants_2 import PARALLEL_SIZE
from collections import deque

LOG_INTERVAL = 100
PERFORMANCE_LOG_INTERVAL = 1000

class A3CTrainingThread(object):
  def __init__(self,
               thread_index,
               global_network,
               initial_learning_rate,
               learning_rate_input,
               grad_applier,
               max_global_time_step,
               device,
               global_network_stable):

    self.thread_index = thread_index
    self.learning_rate_input = learning_rate_input
    self.max_global_time_step = max_global_time_step

    if USE_LSTM:
      self.local_network = GameACLSTMNetwork(ACTION_SIZE, thread_index, device)
    else:
      self.local_network = GameACFFNetwork(ACTION_SIZE, thread_index, device)

    self.local_network.prepare_loss(ENTROPY_BETA)

    with tf.device(device):
      var_refs = [v._ref() for v in self.local_network.get_vars()]
      self.gradients = tf.gradients(
        self.local_network.total_loss, var_refs,
        gate_gradients=False,
        aggregation_method=None,
        colocate_gradients_with_ops=False)

    self.apply_gradients = grad_applier.apply_gradients(
      global_network.get_vars(),
      self.gradients )
      
    self.sync = self.local_network.sync_from(global_network)
    
    self.game_state = GameState(113 * thread_index, thread_index = thread_index)
    
    self.local_t = 0

    self.initial_learning_rate = initial_learning_rate

    self.episode_reward = 0

    # variable controling log output
    self.prev_local_t = 0

    self.state_setted_flag = False
    self.sync_from_stable = self.local_network.sync_from(global_network_stable)
    self.now_reward = 0 # same to episode_reward
    self.episode_scores = deque(maxlen = constants.score_average_length)
    self.beyond = False
    self.score_nums = deque(maxlen = constants.score_average_length)

  def max_v(self):
    return max(self.episode_scores) 

  def mean_v(self):
    return sum(self.episode_scores) * 1.0 / len(self.episode_scores)

  def max_num(self):
    return max(self.score_nums)

  def mean_num(self):
    return sum(self.score_nums) * 1.0 / len(self.score_nums)

  def _anneal_learning_rate(self, global_time_step):
    learning_rate = self.initial_learning_rate * (self.max_global_time_step - global_time_step) / self.max_global_time_step
    if learning_rate < 0.0:
      learning_rate = 0.0
    return learning_rate

  def choose_action(self, pi_values):
    return np.random.choice(range(len(pi_values)), p=pi_values)

  def _record_score(self, sess, summary_writer, summary_op, score_input, score, global_t):
    summary_str = sess.run(summary_op, feed_dict={
      score_input: score
    })
    summary_writer.add_summary(summary_str, global_t)
    summary_writer.flush()
    
  def set_start_time(self, start_time):
    self.start_time = start_time

  def _set_initial_state_multi(self, sess, stable_reward):
    if self.state_setted_flag == False:
      sess.run(self.sync_from_stable)
      pre = PARALLEL_SIZE // 2

      if self.thread_index > pre:
        object_reward = stable_reward
      else:
        if stable_reward > 0:
          num_per_stage = pre // stable_reward
          object_reward = self.thread_index // num_per_stage
        else:
          object_reward = stable_reward

      while self.now_reward < object_reward:
        pi_, value_ = self.local_network.run_policy_and_value(sess, self.game_state.s_t)
        action = self.choose_action(pi_)
        self.game_state.process(action)    
        reward = self.game_state.reward
        terminal = self.game_state.terminal
        self.episode_reward += reward
        self.now_reward += np.clip(reward, -1, 1)
        self.game_state.update()
      self.beyond = False
      self.state_setted_flag = True

  def _set_initial_state(self, sess, stable_reward):
    if self.thread_index > PARALLEL_SIZE // 2 and self.state_setted_flag == False:
      sess.run(self.sync_from_stable)
      #print('thead_index = {}, now_reward = {}, stable_reward = {}'.format(self.thread_index, self.now_reward, stable_reward))
      while self.now_reward < stable_reward:
        pi_, value_ = self.local_network.run_policy_and_value(sess, self.game_state.s_t)
        action = self.choose_action(pi_)
        self.game_state.process(action)    
        reward = self.game_state.reward
        terminal = self.game_state.terminal
        self.episode_reward += reward
        self.now_reward += np.clip(reward, -1, 1)
        self.game_state.update()
      self.beyond = False
      self.state_setted_flag = True
      #print('thread_index = {}, state is setted.'.format(self.thread_index))

  def process(self, sess, global_t, summary_writer, summary_op, score_input, stable_reward):

#    self._set_initial_state(sess, stable_reward)
    self._set_initial_state_multi(sess, stable_reward)

    states = []
    actions = []
    rewards = []
    values = []

    terminal_end = False

    # copy weights from shared to local
    sess.run( self.sync )

    start_local_t = self.local_t

    if USE_LSTM:
      start_lstm_state = self.local_network.lstm_state_out
    
    # t_max times loop
    for i in range(LOCAL_T_MAX):
      pi_, value_ = self.local_network.run_policy_and_value(sess, self.game_state.s_t)
      action = self.choose_action(pi_)

      states.append(self.game_state.s_t)
      actions.append(action)
      values.append(value_)

      if (self.thread_index == 0) and (self.local_t % LOG_INTERVAL == 0):
        print("pi={}".format(pi_))
        print(" V={}".format(value_))

      # process game
      self.game_state.process(action)

      # receive game result
      reward = self.game_state.reward
      terminal = self.game_state.terminal

      self.episode_reward += reward
      self.now_reward += np.clip(reward, -1, 1)

      #####################################################   
      if constants.psc_use:
        reward += self.game_state.psc_reward
      
      #print('psc_reward = {}'.format(self.game_state.psc_reward))
      ######################################################

      # clip reward
      rewards.append( np.clip(reward, -1, 1) )

      self.local_t += 1

      # s_t1 -> s_t
      self.game_state.update()

      #########################################
      if self.now_reward > stable_reward:
        print('thrad_index = {}, beyond the stable_reward. now_reward = {}, stable_reward = {}'.format(self.thread_index, self.now_reward, stable_reward))
        self.beyond = True
#        break
      #######################################

      if terminal or self.beyond:
        self.episode_scores.append(self.episode_reward)
        self.score_nums.append(self.now_reward)

        terminal_end = True
        print("score={}, thread_index = {}, steps={}".format(self.episode_reward, self.thread_index, self.local_t))

        self._record_score(sess, summary_writer, summary_op, score_input,
                           self.episode_reward, global_t)
          
        self.episode_reward = 0
        self.game_state.reset()
        if USE_LSTM:
          self.local_network.reset_state()

        self.beyond = False
        self.state_setted_flag = False
        if self.thread_index > 0:
          self.now_reward = 0
        break

    R = 0.0
    if not terminal_end:
      R = self.local_network.run_value(sess, self.game_state.s_t)
    
    if self.beyond:
      R = self.local_network.run_value(sess, self.game_state.s_t)

    actions.reverse()
    states.reverse()
    rewards.reverse()
    values.reverse()

    batch_si = []
    batch_a = []
    batch_td = []
    batch_R = []

    # compute and accmulate gradients
    for(ai, ri, si, Vi) in zip(actions, rewards, states, values):
      R = ri + GAMMA * R
      td = R - Vi
      a = np.zeros([ACTION_SIZE])
      a[ai] = 1

      batch_si.append(si)
      batch_a.append(a)
      batch_td.append(td)
      batch_R.append(R)

    cur_learning_rate = self._anneal_learning_rate(global_t)

    if USE_LSTM:
      batch_si.reverse()
      batch_a.reverse()
      batch_td.reverse()
      batch_R.reverse()

      sess.run( self.apply_gradients,
                feed_dict = {
                  self.local_network.s: batch_si,
                  self.local_network.a: batch_a,
                  self.local_network.td: batch_td,
                  self.local_network.r: batch_R,
                  self.local_network.initial_lstm_state: start_lstm_state,
                  self.local_network.step_size : [len(batch_a)],
                  self.learning_rate_input: cur_learning_rate } )
    else:
      sess.run( self.apply_gradients,
                feed_dict = {
                  self.local_network.s: batch_si,
                  self.local_network.a: batch_a,
                  self.local_network.td: batch_td,
                  self.local_network.r: batch_R,
                  self.learning_rate_input: cur_learning_rate} )
      
    if (self.thread_index == 0) and (self.local_t - self.prev_local_t >= PERFORMANCE_LOG_INTERVAL):
      self.prev_local_t += PERFORMANCE_LOG_INTERVAL
      elapsed_time = time.time() - self.start_time
      steps_per_sec = global_t / elapsed_time
      print("### Performance : global_t = {} STEPS in {:.0f} seconds. {:.0f} STEPS/sec. {:.2f}M STEPS/hour".format(
        global_t,  elapsed_time, steps_per_sec, steps_per_sec * 3600 / 1000000.))

    # return advanced local step size
    diff_local_t = self.local_t - start_local_t
    return diff_local_t
    
a3c_2.py
# -*- coding: utf-8 -*-
import tensorflow as tf
import threading
import numpy as np

import signal
import random
import math
import os
import time

from game_ac_network_2 import GameACFFNetwork, GameACLSTMNetwork
from a3c_training_thread_2 import A3CTrainingThread
from rmsprop_applier import RMSPropApplier

from constants_2 import ACTION_SIZE
from constants_2 import PARALLEL_SIZE
from constants_2 import INITIAL_ALPHA_LOW
from constants_2 import INITIAL_ALPHA_HIGH
from constants_2 import INITIAL_ALPHA_LOG_RATE
from constants_2 import MAX_TIME_STEP
from constants_2 import CHECKPOINT_DIR
from constants_2 import LOG_FILE
from constants_2 import RMSP_EPSILON
from constants_2 import RMSP_ALPHA
from constants_2 import GRAD_NORM_CLIP
from constants_2 import USE_GPU
from constants_2 import USE_LSTM
from constants_2 import GAME_ENV
import constants_2 as constants

print('action_size = {}'.format(ACTION_SIZE))
print('**************************************')

def log_uniform(lo, hi, rate):
  log_lo = math.log(lo)
  log_hi = math.log(hi)
  v = log_lo * (1-rate) + log_hi * rate
  return math.exp(v)

device = "/cpu:0"
if USE_GPU:
  device = "/gpu:0"

initial_learning_rate = log_uniform(INITIAL_ALPHA_LOW,
                                    INITIAL_ALPHA_HIGH,
                                    INITIAL_ALPHA_LOG_RATE)

global_t = 0

stop_requested = False

if USE_LSTM:
  global_network = GameACLSTMNetwork(ACTION_SIZE, -1, device)
  global_network_stable = GameACLSTMNetwork(ACTION_SIZE, -2, device)
  update_stable_network = global_network_stable.sync_from(global_network)
  stable_reward = 0
else:
  global_network = GameACFFNetwork(ACTION_SIZE, -1, device)
  # add stable_network
  global_network_stable = GameACFFNetwork(ACTION_SIZE, -2, device)
  update_stable_network = global_network_stable.sync_from(global_network)
  stable_reward = 0

training_threads = []

learning_rate_input = tf.placeholder("float")

grad_applier = RMSPropApplier(learning_rate = learning_rate_input,
                              decay = RMSP_ALPHA,
                              momentum = 0.0,
                              epsilon = RMSP_EPSILON,
                              clip_norm = GRAD_NORM_CLIP,
                              device = device)

for i in range(PARALLEL_SIZE):
  training_thread = A3CTrainingThread(i, global_network, initial_learning_rate,
                                      learning_rate_input,
                                      grad_applier, MAX_TIME_STEP,
                                      device = device,
                                      global_network_stable = global_network_stable)
  training_threads.append(training_thread)

# prepare session
sess = tf.Session(config=tf.ConfigProto(log_device_placement=False,
                                        allow_soft_placement=True))

init = tf.global_variables_initializer()
sess.run(init)

# summary for tensorboard
score_input = tf.placeholder(tf.int32)
tf.summary.scalar("score", score_input)

summary_op = tf.summary.merge_all()
summary_writer = tf.summary.FileWriter(LOG_FILE, sess.graph)

# init or load checkpoint with saver
saver = tf.train.Saver()
checkpoint = tf.train.get_checkpoint_state(CHECKPOINT_DIR)
if checkpoint and checkpoint.model_checkpoint_path:
  saver.restore(sess, checkpoint.model_checkpoint_path)
  print("checkpoint loaded:", checkpoint.model_checkpoint_path)
  tokens = checkpoint.model_checkpoint_path.split("-")
  # set global step
  global_t = int(tokens[1])
  print(">>> global step set: ", global_t)
  # set wall time
  wall_t_fname = CHECKPOINT_DIR + '/' + 'wall_t.' + str(global_t)
  with open(wall_t_fname, 'r') as f:
    wall_t = float(f.read())
else:
  print("Could not find old checkpoint")
  # set wall time
  wall_t = 0.0

############################
'''
print(sess.run(global_network.W_conv1)[:,:,0,0])
print(sess.run(global_network_stable.W_conv1)[:,:,0,0])
print('stable_network == global_network: {}'.format(global_network_stable.is_same(global_network)))
sess.run(update_stable_network)
print('stable_network == global_network: {}'.format(global_network_stable.is_same(global_network)))
print(sess.run(global_network_stable.W_conv1)[:,:,0,0])
'''
#############################

def train_function(parallel_index):
  global global_t
  global stable_reward
  global update_stable_network

  training_thread = training_threads[parallel_index]
  # set start_time
  start_time = time.time() - wall_t
  training_thread.set_start_time(start_time)

  while True:
    if stop_requested:
      break
    if global_t > MAX_TIME_STEP:
      break

    diff_global_t = training_thread.process(sess, global_t, summary_writer, summary_op, score_input, stable_reward)
    
    ########################################################################
    if training_thread.thread_index == 0 and training_thread.now_reward > stable_reward:
      print('thead_index = {}, now_reward = {}, stable_reward = {}'.format(training_thread.thread_index, training_thread.now_reward, stable_reward))
      if len(training_thread.episode_scores) == constants.score_average_length:
        max_v = training_thread.episode_socres.max_v()
        mean_v = training_thread.episode_scores.mean_v()
        print('thead_index = {}, q_len = {}, max_v = {}, mean_v ={}'.format(taining_thread.thread_index, len(training_thread.episode_scores, max_v, mean_v)))
        '''
        if mean_v * 1.0 / max_v > 0.8:
          stable_reward = training_thread.now_reward
          sess.run(update_stable_network)
          print('....synchronize the globle_network to stable_network.')
      	'''
      len_score_nums = len(training_thread.score_nums)
      if len_score_nums == constants.score_average_length:
      	max_num = training_thread.score_nums.max_num()
      	mean_num = training_thread.score_nums.mean_num()
      	print('thead_index = {}, q_len = {}, max_num = {}, mean_num ={}'.format(taining_thread.thread_index, len(training_thread.score_nums, max_num, mean_num)))
      	if mean_num * 1.0 / max_num > 0.8:
      		stable_reward = training_thread.now_reward
      		sess.run(update_stable_network)
      		print('....synchronize the globle_network to stable_network.')

      training_thread.now_reward = 0
      ######################################################################
    
    '''
    if global_t % 5 == 0:
      sess.run(update_stable_network)
      print('.............Test: synchronize the globle_network to stable_network.')
    '''

    global_t += diff_global_t
    
    
def signal_handler(signal, frame):
  global stop_requested
  print('You pressed Ctrl+C!')
  stop_requested = True

train_threads = []
for i in range(PARALLEL_SIZE):
  train_threads.append(threading.Thread(target=train_function, args=(i,)))
  
signal.signal(signal.SIGINT, signal_handler)

# set start time
start_time = time.time() - wall_t

for t in train_threads:
  t.start()

print('Press Ctrl+C to stop')
signal.pause()

print('Now saving data. Please wait')
  
for t in train_threads:
  t.join()

if not os.path.exists(CHECKPOINT_DIR):
  os.mkdir(CHECKPOINT_DIR)  

# write wall time
wall_t = time.time() - start_time
wall_t_fname = CHECKPOINT_DIR + '/' + 'wall_t.' + str(global_t)
with open(wall_t_fname, 'w') as f:
  f.write(str(wall_t))

saver.save(sess, CHECKPOINT_DIR + '/' + 'checkpoint', global_step = global_t)
